src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -121,8 +121,19 @@ namespace eicrecon {
           break;
         }
 
-        // 3. find mchit's MCParticle
-        const auto& mcp = mchit->getContributions(0).getParticle();
+        // 3. find mchit's highest energy MCParticle
+        double eMcPar = 0.;
+        double eMcMax = 0.;
+        uint32_t iMcMax = 0;
+        for (uint32_t iContrib = 0; const auto& contrib : mchit->getContributions()) {
+          eMcPar = contrib.getParticle().getEnergy();
+          if (eMcPar > eMcMax) {
```

Comment:
You have more info that you can use here: you know that the total energy of all contributions is the total cluster energy. As you loop over the contributions, you can add up how much of the energy you have already accounted for. E.g. if your cluster has total energy of 5 GeV, your current maximum is 1 GeV, and your running sum is 4.2 GeV (which includes the 1 GeV), then you can stop evaluating the rest of the array. Requires non-log-weighted sum. This may be particularly useful if the most energetic contributions happen to appear early.

Reply:
Oh! Very good point!!